src/events.rs

```diff
@@ -0,0 +1,319 @@
+        let inner: notify::RecommendedWatcher = notify::Watcher::new_immediate(move |res| {
+            // The only way `send` can fail is if the receiver is dropped,
+            // and `MuxedEvents` controls both. `unwrap` is not used,
+            // however, since `Drop` idiosyncrasies could otherwise result
+            // in a panic.
+            let _ = tx.send(res);
+        })
+        .expect("Failed to start watcher");
+
+        MuxedEvents {
+            inner,
+            watched_directories: HashMap::new(),
+            watched_files: HashSet::new(),
+            pending_watched_files: HashSet::new(),
+            event_stream: rx,
+        }
+    }
+
+    fn watch_exists(&self, path: impl AsRef<Path>) -> bool {
+        let path = path.as_ref();
+
+        // Make sure we aren't already watching the directory
+        self.watched_files.contains(&path.to_path_buf())
+            || self.pending_watched_files.contains(&path.to_path_buf())
+            || self.watched_directories.contains_key(&path.to_path_buf())
+    }
+
+    fn add_directory(&mut self, path: impl AsRef<Path> + Into<PathBuf>) -> Result<(), Error> {
+        let path_ref = path.as_ref();
+
+        // TODO: Check the count, but this is okay to call multiple times anyway
+        notify::Watcher::watch(
+            &mut self.inner,
+            &path_ref,
+            notify::RecursiveMode::NonRecursive,
+        )
+        .map_err(|_e| Error::AddFailure)?;
+
+        let count = self.watched_directories.entry(path.into()).or_insert(0);
+        *count += 1;
+
+        Ok(())
+    }
+
+    fn remove_directory(&mut self, path: impl AsRef<Path>) -> Result<(), Error> {
+        let path_ref = path.as_ref();
+
+        if let Some(count) = self.watched_directories.get(path_ref).copied() {
+            match count {
+                0 => unreachable!(), // watch is removed if count == 1
+                1 => {
+                    let _ = notify::Watcher::unwatch(&mut self.inner, &path_ref)
+                        .map_err(|_e| Error::RemoveFailure);
+                    self.watched_directories.remove(path_ref);
+                }
+                _ => {
+                    let new_count = self.watched_directories.get_mut(path_ref).unwrap();
+                    *new_count -= 1;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Adds a given file to the event watch, allowing for files which do not
+    /// yet exist.
+    ///
+    /// Returns the canonicalized version of the path originally supplied, to
+    /// match against the one contained in each `notify::Event` received.
+    /// Otherwise returns `Error` for a given registration failure.
+    pub fn add_file(&mut self, path: impl AsRef<Path> + Into<PathBuf>) -> Result<PathBuf, Error> {
+        let path = absolutify(path, true);
+
+        // Make sure we aren't already watching the directory
+        if self.watch_exists(&path) {
```

Comment:
It doesn't seem like this needs to be an error - it could just early-return with a copy of the canonical path.